File: nginx-lua/lua/s3.lua
-- s3.lua — S3 client using resty.http + AWS SigV4.

local http    = require "resty.http"
local cjson   = require "cjson.safe"
local aws_sig = require "aws_sig"

local _M = {}
local mt = { __index = _M }

function _M.new(cfg)
    local self = {
        endpoint          = cfg.s3_endpoint or "",
        external_endpoint = cfg.s3_external_endpoint or "",
        region            = cfg.aws_region or "us-east-1",
        access_key        = cfg.aws_access_key_id or "",
        secret_key        = cfg.aws_secret_access_key or "",
    }
    if self.external_endpoint == "" then
        self.external_endpoint = self.endpoint
    end
    return setmetatable(self, mt)
end

-- ---------------------------------------------------------------------------
-- Internal helpers
-- ---------------------------------------------------------------------------

-- Build the base URL for S3 requests (path-style).
local function base_url(self)
    if self.endpoint ~= "" then
        return "http://" .. self.endpoint
    end
    return "https://s3." .. self.region .. ".amazonaws.com"
end

-- Make a signed S3 request.
-- Returns body (string), status (number), response headers.
local function s3_request(self, method, bucket, key, body, extra_headers, query)
    local path = "/" .. bucket
    if key and key ~= "" then
        path = path .. "/" .. key
    end

    local host = self.endpoint
    if host == "" then
        host = "s3." .. self.region .. ".amazonaws.com"
    end

    local headers = {
        ["Host"] = host,
    }
    if extra_headers then
        for k, v in pairs(extra_headers) do
            headers[k] = v
        end
    end

    local sign_headers = aws_sig.sign_request({
        method     = method,
        path       = path,
        query      = query or "",
        headers    = headers,
        body       = body or "",
        region     = self.region,
        service    = "s3",
        access_key = self.access_key,
        secret_key = self.secret_key,
    })

    for k, v in pairs(sign_headers) do
        headers[k] = v
    end

    local url = base_url(self) .. path
    if query and query ~= "" then
        url = url .. "?" .. query
    end

    local httpc = http.new()
    httpc:set_timeout(10000) -- 10 seconds

    local res, err = httpc:request_uri(url, {
        method  = method,
        headers = headers,
        body    = body,
    })

    if not res then
        return nil, 0, nil, err
    end

    return res.body, res.status, res.headers, nil
end

-- ---------------------------------------------------------------------------
-- Parse S3 XML responses (minimal parsers for ListObjectsV2 and errors)
-- ---------------------------------------------------------------------------

local function xml_text(xml, tag)
    return xml:match("<" .. tag .. ">(.-)</" .. tag .. ">")
end

local function xml_iter(xml, tag)
    local pos = 1
    return function()
        local s, e = xml:find("<" .. tag .. ">(.-)</" .. tag .. ">", pos)
        if not s then return nil end
        local content = xml:sub(s, e)
        pos = e + 1
        return content
    end
end

-- ---------------------------------------------------------------------------
-- Public API
-- ---------------------------------------------------------------------------

--- List all objects in a bucket. Returns array of {key, size, last_modified}.
function _M.list_objects(self, bucket)
    local result = {}
    local continuation_token = nil

    while true do
        local query = "list-type=2&max-keys=1000"
        if continuation_token then
            query = query .. "&continuation-token=" .. ngx.escape_uri(continuation_token)
        end

        local body, status, _, err = s3_request(self, "GET", bucket, "", nil, nil, query)
        if err or status ~= 200 then
            ngx.log(ngx.ERR, "S3 ListObjects failed for ", bucket, ": ",
                    err or ("HTTP " .. tostring(status) .. " " .. (body or "")))
            break
        end

        for content in xml_iter(body, "Contents") do
            local key  = xml_text(content, "Key")
            local size = tonumber(xml_text(content, "Size") or "0")
            local lm   = xml_text(content, "LastModified") or ""
            if key then
                result[#result + 1] = {
                    key           = key,
                    size          = size,
                    last_modified = lm,
                }
            end
        end

        local is_truncated = xml_text(body, "IsTruncated")
        if is_truncated == "true" then
            continuation_token = xml_text(body, "NextContinuationToken")
            if not continuation_token then break end
        else
            break
        end
    end

    return result
end

--- HEAD an object. Returns {key, size, last_modified} or nil.
function _M.head_object(self, bucket, key)
    local _, status, resp_headers, err = s3_request(self, "HEAD", bucket, key)
    if err or status ~= 200 then
        return nil
    end

    local size = tonumber(resp_headers["Content-Length"] or resp_headers["content-length"] or "0")
    local lm   = resp_headers["Last-Modified"] or resp_headers["last-modified"] or ""

    return {
        key           = key,
        size          = size,
        last_modified = lm,
    }
end

--- Check if an object exists.
function _M.object_exists(self, bucket, key)
    return self:head_object(bucket, key) ~= nil
end

--- GET an object's content. Returns string or nil, error.
function _M.get_object(self, bucket, key)
    local body, status, _, err = s3_request(self, "GET", bucket, key)
    if err or status ~= 200 then
        return nil, err or ("HTTP " .. tostring(status))
    end
    return body
end

--- Try to GET an object's content. Returns content, found (bool).
function _M.try_get_object(self, bucket, key)
    local body, status = s3_request(self, "GET", bucket, key)
    if status == 200 and body then
        return body, true
    end
    return "", false
end

--- PUT an object.
function _M.put_object(self, bucket, key, content, content_type)
    content_type = content_type or "application/octet-stream"
    local extra = {
        ["Content-Type"]   = content_type,
        ["Content-Length"]  = tostring(#content),
    }
    local _, status, _, err = s3_request(self, "PUT", bucket, key, content, extra)
    if err or (status ~= 200 and status ~= 204) then
        ngx.log(ngx.ERR, "S3 PutObject failed: ", err or ("HTTP " .. tostring(status)))
        return false
    end
    return true
end

--- Generate a public URL for an object.
function _M.public_url(self, bucket, key)
    local ep = self.external_endpoint
    if ep ~= "" then
        return "http://" .. ep .. "/" .. bucket .. "/" .. key
    end
    return "https://" .. bucket .. ".s3." .. self.region .. ".amazonaws.com/" .. key
end

--- Generate a presigned URL for an object.
function _M.presigned_url(self, bucket, key, expires)
    expires = expires or 3600
    local host = self.external_endpoint
    if host == "" then
        host = bucket .. ".s3." .. self.region .. ".amazonaws.com"
    end
    local path = "/" .. bucket .. "/" .. key
    local scheme = "http"
    if self.endpoint == "" then
        scheme = "https"
        path = "/" .. key
    end

    return aws_sig.presigned_url({
        host       = host,
        path       = path,
        region     = self.region,
        service    = "s3",
        access_key = self.access_key,
        secret_key = self.secret_key,
        expires    = expires,
        scheme     = scheme,
    })
end

return _M

File: golang/internal/objectstore/s3.go
package objectstore

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/s3/types"

	"github.com/rachlenko/parparchik/golang/internal/config"
)

// S3Store implements Store against any S3-compatible endpoint (AWS S3 or
// MinIO), using aws-sdk-go-v2 for request signing instead of a hand-rolled
// SigV4 implementation.
type S3Store struct {
	client           *s3.Client
	presign          *s3.PresignClient
	region           string
	externalEndpoint string
}

// NewS3Store builds an S3Store from parparchik configuration. When
// cfg.S3Endpoint is set, requests target that endpoint path-style (the MinIO
// case); otherwise the SDK's default AWS endpoint resolution for cfg.Region
// applies.
func NewS3Store(ctx context.Context, cfg *config.Config) (*S3Store, error) {
	loadOpts := []func(*awsconfig.LoadOptions) error{
		awsconfig.WithRegion(cfg.AWSRegion),
	}
	if cfg.AWSAccessKeyID != "" && cfg.AWSSecretAccessKey != "" {
		loadOpts = append(loadOpts, awsconfig.WithCredentialsProvider(
			credentials.NewStaticCredentialsProvider(cfg.AWSAccessKeyID, cfg.AWSSecretAccessKey, ""),
		))
	}

	awsCfg, err := awsconfig.LoadDefaultConfig(ctx, loadOpts...)
	if err != nil {
		return nil, fmt.Errorf("objectstore: load AWS config: %w", err)
	}

	// Resolved once here (not per-request in PublicURL) so a bare
	// "host:port" only logs its http:// assumption warning once, and so
	// PublicURL never has to re-parse or re-guess the scheme.
	external := cfg.S3ExternalEndpoint
	if external == "" {
		external = cfg.S3Endpoint
	}
	var externalURL string
	if external != "" {
		externalURL = resolveEndpointURL(external)
	}

	client := s3.NewFromConfig(awsCfg, func(o *s3.Options) {
		if cfg.S3Endpoint != "" {
			o.BaseEndpoint = aws.String(resolveEndpointURL(cfg.S3Endpoint))
			o.UsePathStyle = true
		}
	})

	// Presigned URLs must be signed against — and therefore must use —
	// whatever host the *client that follows the redirect* will actually
	// connect to, not necessarily the same host this process uses to talk
	// to storage directly. In the common docker-compose shape,
	// S3_ENDPOINT=minio:9000 (only resolvable inside the Docker network)
	// while S3_EXTERNAL_ENDPOINT=localhost:9000 (what a client outside the
	// network must use); reusing the internal client to presign would
	// produce a URL an external client can't even resolve. When no custom
	// endpoint is configured at all (real AWS S3), presigning against the
	// default client is correct as-is — there's nothing to redirect away
	// from.
	presignClient := client
	if cfg.S3Endpoint != "" {
		presignClient = s3.NewFromConfig(awsCfg, func(o *s3.Options) {
			o.BaseEndpoint = aws.String(externalURL)
			o.UsePathStyle = true
		})
	}

	return &S3Store{
		client:           client,
		presign:          s3.NewPresignClient(presignClient),
		region:           cfg.AWSRegion,
		externalEndpoint: externalURL,
	}, nil
}

func (s *S3Store) ListObjects(ctx context.Context, bucket string) ([]Object, error) {
	var result []Object

	paginator := s3.NewListObjectsV2Paginator(s.client, &s3.ListObjectsV2Input{
		Bucket: aws.String(bucket),
	})
	for paginator.HasMorePages() {
		page, err := paginator.NextPage(ctx)
		if err != nil {
			return nil, fmt.Errorf("objectstore: list objects in %q: %w", bucket, err)
		}
		for _, obj := range page.Contents {
			result = append(result, Object{
				Key:          aws.ToString(obj.Key),
				Size:         aws.ToInt64(obj.Size),
				LastModified: formatTime(obj.LastModified),
			})
		}
	}

	return result, nil
}

func (s *S3Store) HeadObject(ctx context.Context, bucket, key string) (*Object, error) {
	out, err := s.client.HeadObject(ctx, &s3.HeadObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		if isNotFound(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("objectstore: head %q/%q: %w", bucket, key, err)
	}

	return &Object{
		Key:          key,
		Size:         aws.ToInt64(out.ContentLength),
		LastModified: formatTime(out.LastModified),
	}, nil
}

func (s *S3Store) GetObject(ctx context.Context, bucket, key string) ([]byte, error) {
	out, err := s.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		if isNotFound(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("objectstore: get %q/%q: %w", bucket, key, err)
	}
	defer out.Body.Close()

	body, err := io.ReadAll(out.Body)
	if err != nil {
		return nil, fmt.Errorf("objectstore: read %q/%q body: %w", bucket, key, err)
	}
	return body, nil
}

func (s *S3Store) PutObject(ctx context.Context, bucket, key string, content []byte, contentType string) error {
	if contentType == "" {
		contentType = "application/octet-stream"
	}
	_, err := s.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(bucket),
		Key:         aws.String(key),
		Body:        bytes.NewReader(content),
		ContentType: aws.String(contentType),
	})
	if err != nil {
		return fmt.Errorf("objectstore: put %q/%q: %w", bucket, key, err)
	}
	return nil
}

func (s *S3Store) PublicURL(bucket, key string) string {
	if s.externalEndpoint != "" {
		// externalEndpoint is already a full scheme://host URL (resolved
		// once in NewS3Store via resolveEndpointURL) — do not re-prepend a
		// scheme here, or a caller-supplied "https://cdn.example.com"
		// becomes the malformed "http://https://cdn.example.com/...".
		return s.externalEndpoint + "/" + bucket + "/" + key
	}
	return fmt.Sprintf("https://%s.s3.%s.amazonaws.com/%s", bucket, s.region, key)
}

func (s *S3Store) PresignedURL(ctx context.Context, bucket, key string, expires time.Duration) (string, error) {
	req, err := s.presign.PresignGetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	}, s3.WithPresignExpires(expires))
	if err != nil {
		return "", fmt.Errorf("objectstore: presign %q/%q: %w", bucket, key, err)
	}
	return req.URL, nil
}

// resolveEndpointURL accepts either a bare "host:port" (assumed http://, for
// local/internal deployments like the docker-compose MinIO stack — same
// assumption the Lua original hardcoded) or a full "scheme://host:port"
// URL, in which case the caller's scheme (e.g. https://) is honored as-is.
// The Lua original always hardcoded http:// with no way to opt into TLS for
// the internal S3 endpoint at all — see the nginx-lua security review's
// MEDIUM finding on this.
func resolveEndpointURL(endpoint string) string {
	if strings.Contains(endpoint, "://") {
		return endpoint
	}
	slog.Warn("objectstore: S3_ENDPOINT has no scheme, assuming plaintext http:// — set a full https:// URL for any non-local deployment", "endpoint", endpoint)
	return "http://" + endpoint
}

func isNotFound(err error) bool {
	var nf *types.NotFound
	var nsk *types.NoSuchKey
	return errors.As(err, &nf) || errors.As(err, &nsk)
}

func formatTime(t *time.Time) string {
	if t == nil {
		return ""
	}
	return t.UTC().Format(time.RFC3339)
}
